Fix startup issues of Tor

Tor setup sometimes hangs specially at first time using Bisq2.

Hi Guys,

it is my first time running the application and I think i get the issue you are presenting here.

Looking at the logs, I get this:


`mar.-17 15:50:59.212 [NetworkService.network-IO-pool-0] ERROR b.d.DesktopApplicationService: Initializing applicationService failed java.util.concurrent.CompletionException: bisq.tor.controller.exceptions.TorBootstrapFailedException: Tor bootstrap timout triggered.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: bisq.tor.controller.exceptions.TorBootstrapFailedException: Tor bootstrap timout triggered.
	at bisq.tor.controller.NativeTorController.waitUntilBootstrapped(NativeTorController.java:146)
	at bisq.tor.TorService.initialize(TorService.java:99)
	at bisq.network.p2p.node.transport.TorTransportService.initialize(TorTransportService.java:63)
	at bisq.network.p2p.ServiceNode.getInitializedDefaultNode(ServiceNode.java:229)
	at bisq.network.p2p.ServiceNodesByTransport.lambda$getInitializedDefaultNodeByTransport$1(ServiceNodesByTransport.java:133)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	... 3 common frames omitted`
	
 Another thing that I notice, every time that I run the application it goes a bit further in the Bootstrap %.

first stoped with 3%, then 6%, 12.50%, 12.75%

Hope that it helps.

--- a/network/tor/tor/src/main/java/bisq/tor/controller/NativeTorController.java
+++ b/network/tor/tor/src/main/java/bisq/tor/controller/NativeTorController.java
@@ -34,6 +34,8 @@ import net.freehaven.tor.control.TorControlConnection;
 
 import java.io.IOException;
 import java.net.Socket;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
@@ -139,11 +141,14 @@ public class NativeTorController implements BootstrapEventListener, HsDescUpload
 
     public void waitUntilBootstrapped() {
         try {
-            boolean isSuccess = isBootstrappedCountdownLatch.await(bootstrapTimeout, TimeUnit.MILLISECONDS);
-            if (isSuccess) {
-                removeBootstrapEventListener();
-            } else {
-                throw new TorBootstrapFailedException("Tor bootstrap timeout triggered.");
+            while (true) {
+                boolean isSuccess = isBootstrappedCountdownLatch.await(bootstrapTimeout, TimeUnit.MILLISECONDS);
+                if (isSuccess) {
+                    removeBootstrapEventListener();
+                    break;
+                } else if (isBootstrapTimeoutTriggered()) {
+                    throw new TorBootstrapFailedException("Tor bootstrap timeout triggered.");
+                }
             }
         } catch (InterruptedException e) {
             throw new TorBootstrapFailedException(e);
@@ -220,5 +225,12 @@ public class NativeTorController implements BootstrapEventListener, HsDescUpload
         TorControlConnection controlConnection = torControlConnection.orElseThrow();
         controllerEventHandler.removeHsDescUploadedListener(this);
         clearAllEventSubscriptionsOnConnection(controlConnection);
+    }
+
+    private boolean isBootstrapTimeoutTriggered() {
+        BootstrapEvent bootstrapEvent = this.bootstrapEvent.get();
+        Instant timestamp = bootstrapEvent.getTimestamp();
+        Instant bootstrapTimeoutAgo = Instant.now().minus(bootstrapTimeout, ChronoUnit.MILLIS);
+        return bootstrapTimeoutAgo.isAfter(timestamp);
     }
 }

--- a/network/tor/tor/src/main/java/bisq/tor/controller/events/events/BootstrapEvent.java
+++ b/network/tor/tor/src/main/java/bisq/tor/controller/events/events/BootstrapEvent.java
@@ -21,6 +21,8 @@ import lombok.Builder;
 import lombok.Getter;
 import lombok.ToString;
 
+import java.time.Instant;
+
 @Builder
 @Getter
 @ToString
@@ -31,6 +33,8 @@ public class BootstrapEvent {
     private final int progress;
     private final String tag;
     private final String summary;
+
+    private final Instant timestamp = Instant.now();
 
     public BootstrapEvent(int progress, String tag, String summary) {
         if (progress < 0 || tag.isEmpty() || summary.isEmpty()) {

